contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,213 @@
+    uint256 public currentReserveImbalance;
+    uint256 public totalMinted = 0;
+
+    // Used for sending and receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address nativeTokenSourceAddress_,
+        uint256 initialReserveImbalance_
+    ) {
+        require(
+            teleporterMessengerAddress != address(0),
+            "NativeTokenDestination: zero TeleporterMessenger address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            sourceBlockchainID_ != bytes32(0),
+            "NativeTokenDestination: zero source chain ID"
+        );
+        require(
+            sourceBlockchainID_ !=
+                IWarpMessenger(0x0200000000000000000000000000000000000005)
+                    .getBlockchainID(),
+            "NativeTokenDestination: cannot bridge with same blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            nativeTokenSourceAddress_ != address(0),
+            "NativeTokenDestination: zero source contract address"
+        );
+        nativeTokenSourceAddress = nativeTokenSourceAddress_;
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "NativeTokenDestination: unauthorized TeleporterMessenger contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "NativeTokenDestination: invalid source chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            senderAddress == nativeTokenSourceAddress,
+            "NativeTokenDestination: unauthorized sender"
+        );
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(
+            recipient != address(0),
+            "NativeTokenDestination: zero recipient address"
+        );
+        require(amount != 0, "NativeTokenDestination: zero transfer value");
+
+        uint256 adjustedAmount = amount;
+        if (currentReserveImbalance > 0) {
+            if (amount > currentReserveImbalance) {
+                emit CollateralAdded({
```

Comment:
should we add a field in `CollateralAdded` about who sent the collateral? In this case `recipient`

Reply:
Do you think it makes more sense to label it with `senderAddress` or `recipient` here? We could also do both. If only one, I think sender makes more sense, because if someone is just sending the transaction to collateralize the bridge, they may enter a dummy recipient address.

Reply:
I think sender makes sense